File: Linked-Lists/876.js
// 876. Middle of the Linked List
/*
Given the head of a singly linked list, return the middle node of the linked list.

If there are two middle nodes, return the second middle node.
*/

// Definition for singly-linked list. 
class ListNode {
    constructor(val, next) {
        this.val = (val===undefined ? 0 : val)
        this.next = (next===undefined ? null : next)
    }
    append(val) {
        const newNode = new ListNode(val)
        let walker = this
        while (walker.next) {
            walker = walker.next
        }
        walker.next = newNode
    }
    printAsArray() {
        let array = []
        let walker = this
        while (walker) {
            array.push(walker.val + ' ' + '->')
            walker = walker.next
        }
        array.push('null')
        console.log(array)
    }
}

var middleNode = function(head) {
    let count = 0
    let walker = head
    while (walker) {
        walker = walker.next
        count++
    }
    let middle = count / 2
    walker = head
    count = 1
    while (count <= middle) {
        walker = walker.next
        count++
    }
    return walker
};


const head = new ListNode(1)
head.append(2)
head.append(3)
head.append(4)
head.append(5)
head.append(6)

let result = middleNode(head)
result.printAsArray()

File: Linked-Lists/876.py
# 876. Middle of the Linked List
# Given the head of a singly linked list, return the middle node of the linked list.
# If there are two middle nodes, return the second middle node.

from typing import Optional

class ListNode:
    def __init__(self, val = 0, next = None):
        self.val = val
        self.next = next

class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        length = 0
        walker = head
        while walker:
            length += 1
            walker = walker.next
        mid = length // 2 + 1
        idx = 1
        walker = head
        while idx != mid:
            walker = walker.next
            idx += 1
        return walker

node5 = ListNode(5)
node4 = ListNode(4, node5)
node3 = ListNode(3, node4)
node2 = ListNode(2, node3)
test_case = ListNode(1, node2)

s = Solution()
midNode = s.middleNode(test_case)

# Print midNode as a string
display = ""
walker = midNode
while walker:
    display += str(walker.val) + " -> "
    walker = walker.next
print(display + "None")